swagger_spec_compatibility/walkers/additional_properties.py

```diff
@@ -39,6 +40,45 @@ def fix_parameter_path(self, path, original_path):
         )
 
 
+CycleSentinel = collections.namedtuple('CycleSentinel', ('path',))
+
+
+def _cycle_safe_dict(dict_with_cycles):
+    # type: (typing.Dict[typing.Any, typing.Any]) -> typing.Dict[typing.Any, typing.Any]
+    """Return a new dict with cycles replaced by sentinels.
+
+    This can be used to do a deep equality comparison against dictionaries
+    which can contain cycles (e.g. expanded Swagger specs).
+    """
+    seen = {}  # type: typing.Dict[int, typing.Tuple[typing.Any, ...]]
+
+    def process(d, path):
+        # type: (typing.Any, typing.Tuple[typing.Any, ...]) -> typing.Any
+        if id(d) in seen:
+            return CycleSentinel(seen[id(d)])
+        else:
+            if isinstance(d, dict):
+                seen[id(d)] = path
+                return {
+                    key: process(value, path + (key,))
+                    for key, value in sorted(d.items())
+                }
+            else:
+                return d
+
+    ret = process(dict_with_cycles, ())
+    assert isinstance(ret, dict)  # mypy isn't smart enough to figure this out
+    return ret
+
+
+def _cycle_safe_compare(a, b):
```

Comment:
I'm wondering if it would be more interesting to move it into a util method or in the base walker class such that other places might be using it.

I'm not convinced that `_evaluate_additional_properties_diffs` is the only method subject to this bug and having it shared would help to address them in a easier way.